include/envoy/http/filter.h

```diff
@@ -119,6 +119,42 @@ enum class FilterMetadataStatus {
   Continue,
 };
 
+/**
+ * An interface for a per-connection object that is created by a filter and then managed by the
+ * HTTP connection manager. This pushes singleton handling into the HTTP connection manager.
+ */
+
+class MutableHttpConnection;
+
+class PerConnectionObject {
+public:
+  virtual ~PerConnectionObject() = default;
+
+  /**
+   * @return MutableHttpConnection& the mutable http connection, this is expected not to be nullptr.
+   */
+  virtual MutableHttpConnection& mutable_http_connection() PURE;
+};
+using PerConnectionObjectSharedPtr = std::shared_ptr<PerConnectionObject>;
+
+/**
+ * An interface for connection attributes that are modifiable by HTTP filters. Generally, these
+ * should be modified only via a PerConnectionObject.
+ */
+class MutableHttpConnection {
+public:
+  virtual ~MutableHttpConnection() = default;
+
+  // Fill with the things that we want HTTP filters to be able to do to the connection such as
+  // setting certain socket options, etc. Everything needs to be abstracted and then flowed through
+  // to the underlying writable Network::Connection.
+
+  /**
+   * @return Network::Connection* the originating connection, this is expected not to be nullptr.
+   */
+  virtual Network::Connection* connection() PURE;
```

Comment:
We cannot expose the raw network connection from this object, only sanitized and controlled methods, so this shoul dbe removed.

Reply:
I just want to get deeper here.

This was my interpretation...

Say one has an instance of MutableHttpConnection *mutable_connection;
In this case, to perform network socket operation, I was thinking of the following abstraction with the use of connection();

so it would look like,
mutable_connection->connection()->setSocketRecvBufferSize(); 'setSocketRecvBufferSize' is an API defined inside class `Network::Connection`.

If this is the not expected; based on the notes shared for MutableHttpConnection; one could use an API like the following as well ?

`MutableHttpConnection::setSocketRecvBufferSize(Network::Connection* conn, uint32_t size)`

If this the intended usage; it makes sense.

I am not fully aware about the design, but when plugged into filter code; from a filter to access connection; we could do `read_callbacks_->connection()` (accessed via ConnectionManagerImpl)

The connection() here is available without being sanitized or is it?

Just curious about the method on exposing a sanitized Network::connection instance ...

Thanks!

Reply:
The point is we do not want to expose the entire non-const Network::Connection to the HTTP filters, only a subset, in a new interface. This can just be directly done by MutableHttpConnection as you propose.